go-sdk/pkg/sdk/keptn.go

```diff
@@ -258,179 +213,117 @@ func (k *Keptn) gotEvent(ctx context.Context, event cloudevents.Event) {
+					if err := eventSender(*finishedEvent); err != nil {
 						k.logger.Errorf("unable to send '.finished' event: %v", err)
 						return
 					}
 				}
 			}
 		}
 	})
+	return nil
 }
 
-func (k *Keptn) runEventTaskAction(fn func()) {
-	if k.syncProcessing {
-		fn()
-	} else {
-		go fn()
-	}
-}
-
-func (k *Keptn) send(event cloudevents.Event) error {
-	k.logger.Infof("Sending %s event", event.Type())
-	if err := k.eventSender.SendEvent(event); err != nil {
-		k.logger.Error("Error sending .started event")
+func (k *Keptn) RegistrationData() controlplane.RegistrationData {
+	subscriptions := []models.EventSubscription{}
+	subjects := []string{}
+	if k.env.PubSubTopic != "" {
+		subjects = strings.Split(k.env.PubSubTopic, ",")
 	}
-	return nil
-}
 
-func (k *Keptn) createStartedEventForTriggeredEvent(triggeredEvent cloudevents.Event) (*cloudevents.Event, error) {
-	startedEventType, err := keptnv2.ReplaceEventTypeKind(triggeredEvent.Type(), "started")
-	if err != nil {
-		return nil, fmt.Errorf("unable to create '.finished' event: %v from %s", err, triggeredEvent.Type())
+	for _, s := range subjects {
+		subscriptions = append(subscriptions, models.EventSubscription{Event: s})
 	}
-	keptnContext, err := triggeredEvent.Context.GetExtension(KeptnContextCEExtension)
-	if err != nil {
-		return nil, fmt.Errorf("unable to get keptn context from '.triggered' event: %v", err)
+	return controlplane.RegistrationData{
+		Name: k.source,
+		MetaData: models.MetaData{
+			Hostname:           k.env.K8sNodeName,
+			IntegrationVersion: k.env.Version,
+			Location:           k.env.Location,
+			DistributorVersion: "0.15.0", // note: to be deleted when bridge stops requiring this info
+			KubernetesMetaData: models.KubernetesMetaData{
+				Namespace:      k.env.K8sNamespace,
+				PodName:        k.env.K8sPodName,
+				DeploymentName: k.env.K8sDeploymentName,
+			},
+		},
+		Subscriptions: subscriptions,
 	}
-	eventData := keptnv2.EventData{}
-	triggeredEvent.DataAs(&eventData)
-	c := cloudevents.NewEvent()
-	c.SetID(uuid.New().String())
-	c.SetType(startedEventType)
-	c.SetDataContentType(cloudevents.ApplicationJSON)
-	c.SetExtension(KeptnContextCEExtension, keptnContext)
-	c.SetExtension(TriggeredIDCEExtension, triggeredEvent.ID())
-	c.SetSource(k.source)
-	c.SetData(cloudevents.ApplicationJSON, eventData)
-	return &c, nil
 }
 
-func (k *Keptn) createFinishedEventForReceivedEvent(receivedEvent cloudevents.Event, eventData interface{}) (*cloudevents.Event, error) {
-	var genericEvent map[string]interface{}
-	keptnv2.Decode(eventData, &genericEvent)
-	if genericEvent["status"] == nil || genericEvent["status"] == "" {
-		genericEvent["status"] = "succeeded"
+func (k *Keptn) Start() error {
+	if k.env.HealthEndpointEnabled {
+		k.healthEndpointRunner(k.env.HealthEndpointPort, k.controlPlane)
 	}
+	ctx := k.getContext(k.gracefulShutdown)
+	err := k.controlPlane.Register(ctx, k)
+	ctx.Value(gracefulShutdownKey).(wgInterface).Wait()
```

Comment:
should we add a check whether the cast to `wgInterface` was succesful here? Alternatively, we could return the waitGroup as a result of `getContext()`

Reply:
good idea, will return the wg